File: today3/search-in-rotated-sorted-array-ii.ts
/**
 * https://leetcode.com/problems/search-in-rotated-sorted-array-ii/description/
 * [NeetCode Solution](https://www.youtube.com/watch?v=oUnF7o88_Xc&t=359s)
 */
export { search };

function search(nums: number[], target: number): boolean {
  function bs(start: number, end: number): boolean {
    const mid = Math.floor((start + end) / 2);
    if (nums[mid] === target) return true;
    if (start >= end) return false;

    if (nums[start] <= nums[mid]) {
      // in left rotation
      if (target > nums[mid]) {
        return bs(mid + 1, end);
      } else {
        if (target >= nums[start]) return bs(start, mid - 1);
        else return bs(mid + 1, end);
      }
    } else {
      // in right rotation
      if (target < nums[mid]) {
        return bs(start, mid - 1);
      } else {
        if (target <= nums[end]) return bs(mid + 1, end);
        else return bs(start, mid - 1);
      }
    }
  }

  return bs(0, nums.length - 1);
}

File: python/today6/search_in_rotated_sorted_array_ii.py
# https://leetcode.com/problems/search-in-rotated-sorted-array-ii/
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums) - 1

        while l <= r:
            m = (l + r) // 2

            if nums[m] == target:
                return True
            
            # left is sorted
            if nums[l] < nums[m]:
                if nums[l] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            # right is sorted
            elif nums[l] > nums[m]:
                if nums[m] < target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1            
            else:
                l += 1
        return False